---
title: "Troubleshooting the cluster similarity matrices"
subtitle: "Assessing gene expression spaces"
author: "Antoine Beauchamp"
date: '2022-06-09'
output: 
  html_document:
    theme: paper
    highlight: pygments
    code_folding: hide
---

# Initialization

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(RMINC))
suppressPackageStartupMessages(library(data.tree))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(viridisLite))
suppressPackageStartupMessages(library(ggplotify))
```

# Functions

```{r}
source('../../functions/tree_tools.R')
source('../../functions/buildSimilarityMatrix.R')

#' Title
#'
#' @param x 
#' @param tree 
#' @param labels 
#' @param mask 
#'
#' @return
label_mouse_data <- function(x, tree, labels, mask) {
  
  labels_pruned <- hanatToAtlas(tree, mincArray(labels))
  defs_pruned <- hanatToAtlasDefs(tree)
  
  labels_pruned <- labels_pruned[mask == 1]
  
  x <- x %>% 
    mutate(label = labels_pruned) %>% 
    left_join(defs_pruned, by = c('label' = 'Label')) %>% 
    select(-label) %>% 
    rename(Region = Structure)
  
  return(x)
  
}

#' Title
#'
#' @param x 
#' @param tree 
#' @param samples 
#'
#' @return
label_human_data <- function(x, tree, samples) {
  
  tree <- Clone(tree)
  
  tree$Do(function(node){
    node$struct <- rep(node$name, length(node$samples))
  }, traversal = 'post-order')
  
  tree_structs <- unlist(tree$Get('struct', filterFun = isLeaf))
  names(tree_structs) <- NULL
  
  tree_samples <- unlist(tree$Get('samples', filterFun = isLeaf))
  names(tree_samples) <- NULL
  
  sample_defs <- tibble(Structure = tree_structs,
                        sample_id = tree_samples)
  
  x <- x %>% 
    mutate(sample_id = samples) %>% 
    left_join(sample_defs, by = 'sample_id') %>% 
    select(-sample_id) %>% 
    rename(Region = Structure)
    
  return(x)
  
}

aggregate <- function(x, groupby, method = 'mean', output = 'tibble'){
  
  if (method == 'mean') {
    aggfunc <- mean
  } else if (method == 'median') {
    aggfunc <- median
  } else if (method == 'sd') {
    aggfunc <- sd
  } else {
    stop()
  }
  
  if (!(groupby %in% colnames(x))) {
    stop()
  }
  
  if (any(is.na(x[[groupby]]))){
    warning(paste("Grouping variable", groupby, "contains NA.",
                  "These observations will be ommitted when aggregating."))
  }
  
  x <- x %>% 
    filter(!is.na(.data[[groupby]])) %>% 
    group_by_at(.vars = vars(groupby)) %>% 
    summarise_all(.funs = aggfunc) %>% 
    ungroup() 
  
  if (output == 'tibble') {
    x %>% 
      as_tibble() %>% 
      return()
  } else if (output == 'data.frame') {
    x %>% 
      as.data.frame(stringsAsFactors = FALSE) %>% 
      return()
  } else if (output == 'matrix') {
    x %>% 
      column_to_rownames(groupby) %>% 
      as.matrix() %>% 
      return()
  } else {
    stop()
  }
}

generate_heatmap <- function(mousefile, humanfile, mousetree, humantree, mouseregions, humanregions, labels, mask, samples) {
  
  mouse <- data.table::fread(mousefile, header = TRUE) %>% 
    as_tibble()
  human <- data.table::fread(humanfile, header = TRUE) %>%
    as_tibble()
  
  mouse <- mouse %>% 
    label_mouse_data(tree = mousetree,
                     labels = labels,
                     mask = mask) %>% 
    filter(!is.na(Region)) %>% 
    aggregate(groupby = 'Region',
              output = 'matrix') %>%
    t()
  
  human <- human %>% 
    label_human_data(tree = humantree,
                     samples = samples) %>% 
    aggregate(groupby = 'Region',
              output = 'matrix') %>% 
    t()
  
  similarity <- buildSimilarityMatrix(x1 = human, 
                                      x2 = mouse,
                                      method = 'correlation')
  
  similarity <- similarity[,match(mouseregions, colnames(similarity))]
  similarity <- similarity[match(humanregions, rownames(similarity)),]
  
  p <- pheatmap(similarity,
                cluster_cols = F,
                cluster_rows = F,
                silent = TRUE, 
                na_col = 'black')
  
  return(p)
}
```

# Importing

```{r}
#Mouse files
mouse_labelfile <- '../../data/mouse/atlas/DSURQE_CCFv3_labels_200um.mnc'
mouse_maskfile <- '../../data/mouse/atlas/coronal_200um_coverage_bin0.8.mnc'
mouse_treefile <- '../../data/mouse/MouseExpressionTree_DSURQE.RData'

#Human files
human_treefile <- '../../data/human/HumanExpressionTree.RData'
human_samplefile <- '../../data/human/SampleInformation_pipeline_v1.csv'

#Tree labels
treelabels <- '../../data/TreeLabelsReordered.RData'

#Trees
load(mouse_treefile)
tree_mouse <- Clone(treeMouseExpr)
rm(treeMouseExpr)

load(human_treefile)
tree_human <- Clone(treeHumanExpr)
rm(treeHumanExpr)

#Tree labels
load(treelabels)

#Prune mouse tree to desired level
mouse_regions <- c(listLabelsMouseReordered$Region67, 
                   'fiber tracts', 
                   'ventricular systems')
tree_mouse_pruned <- Clone(tree_mouse)
pruneAnatTree(tree_mouse_pruned, 
              nodes = mouse_regions, 
              method = 'BelowNode')

#Pruned human tree to desired level
human_regions <- c(listLabelsHumanReordered$Region88, 
                   'white matter', 
                   'sulci & spaces')
tree_human_pruned <- Clone(tree_human)
pruneAnatTree(tree_human_pruned, 
              nodes = human_regions, 
              method = 'BelowNode')

#DSURQE labels and AMBA mask
labels <- mincGetVolume(mouse_labelfile)
mask <- mincGetVolume(mouse_maskfile)

#Human microarray sample IDs
sample_ids <- data.table::fread(human_samplefile, header = TRUE) %>% 
  as_tibble() %>% 
  pull(SampleID)
```

# Overview 

First I want to get some sense of whether I generated the gene expression spaces incorrectly in this project. There are some slight differences in the processing, compared with the transcriptomic similarity project, so I want to make sure that I haven't broken something. To get a sense of this, I'm going to use the voxel-wise gene expression data sets to generate regional similarity matrices at the same level of aggregation as in the previous project. It should be apparent if there is something wrong with the data at this stage. First we take a look at the similarity matrix in the input gene space. 

# Troubleshooting

## Input space similarity matrix

```{r fig.width = 14}
mouse_file <- 'data/MouseExpressionMatrix_voxel_coronal_log2_grouped_imputed_homologs_scaled.csv'
human_file <- 'data/HumanExpressionMatrix_samples_pipeline_v1_homologs_scaled.csv'

p_inputspace <- generate_heatmap(mousefile = mouse_file,
                                 humanfile = human_file,
                                 mousetree = tree_mouse_pruned,
                                 humantree = tree_human_pruned,
                                 mouseregions = mouse_regions,
                                 humanregions = human_regions,
                                 labels = labels,
                                 mask = mask,
                                 samples = sample_ids)

as.ggplot(p_inputspace)
```
This matrix looks the way I expect it to, so it doesn't seem as though there's anything wrong with the voxelwise gene expression data. The one notable difference here, which may play a part is that this data set includes voxels corresponding to white matter and ventricular spaces. I'll ignore this for now but will consider it later. 

## Individual latent spaces

Next I want to determine whether the latent space data is behaving as expected. In the transcriptomic similarity project, I used the perceptron to transform regionally aggregated expression values. The process here is slightly different, in that I transform the full voxel-wise expression matrices. If the code works as intended, I don't expect that difference to have a large impact, but it's worth checking. I'll do so by randomly choosing one of the latent spaces and seeing what the similarity matrix looks like in any given space. 

```{r fig.width = 14}
mouse_latent_space_files <- list.files('data/mouse/expression/latent_space_100/', full.names = T) %>% 
  str_subset('mousetransform') %>% 
  sort()
human_latent_space_files <- list.files('data/human/expression/latent_space_100/', full.names = T) %>% 
  str_subset('humantransform') %>% 
  sort()

nfiles <- length(mouse_latent_space_files)

i <- sample(1:nfiles, size = 1)

p_latentspace_100 <- generate_heatmap(mousefile = mouse_latent_space_files[i],
                                 humanfile = human_latent_space_files[i],
                                 mousetree = tree_mouse_pruned,
                                 humantree = tree_human_pruned,
                                 mouseregions = mouse_regions,
                                 humanregions = human_regions,
                                 labels = labels,
                                 mask = mask,
                                 samples = sample_ids)

as.ggplot(p_latentspace_100)
```

This is what I would expect. Given the random nature of this sampling, each similarity matrix will be slightly different, but they should contain the broad patterns from the input gene space, as well as some pockets of more defined signal. 

One possibility for why the cluster signatures might be different is if I somehow messed up the association between the mouse and human latent spaces. The latent spaces from any given training run differ from one another by an arbitrary rotation, and so it only makes sense to evaluate the similarity of mouse and human data in the same latent space. In that case, the similarity matrix looks like the one above. To get a sense of what a mismatched similarity would look like, we can intentionally use different latent spaces:

```{r fig.width = 14}
i <- sample(1:nfiles, size = 1)
j <- sample(1:nfiles, size = 1)

p_latentspace_100_scrambled <- generate_heatmap(mousefile = mouse_latent_space_files[i],
                                                humanfile = human_latent_space_files[j],
                                                mousetree = tree_mouse_pruned,
                                                humantree = tree_human_pruned,
                                                mouseregions = mouse_regions,
                                                humanregions = human_regions,
                                                labels = labels,
                                                mask = mask,
                                                samples = sample_ids)

as.ggplot(p_latentspace_100_scrambled)
```

As expected, the patterns are random and nonsensical. 

This suggests that the latent spaces were generated as intended and indeed behave as intended. To make sure that this is still true in the case of the average, we can compute the average similarity matrix over 100 latent spaces.

## Average similarity matrix

```{r}
list_sim <- vector(mode = 'list', length = nfiles)
for (i in 1:length(list_sim)){
  
  # if (i %% 10 == 0) {print(i)}
  
  mouse <- data.table::fread(mouse_latent_space_files[i], header = TRUE) %>%
    as_tibble()
  human <- data.table::fread(human_latent_space_files[i], header = TRUE) %>%
    as_tibble()
  
  mouse <- mouse %>%
    label_mouse_data(tree = tree_mouse_pruned,
                     labels = labels,
                     mask = mask) %>%
    filter(!is.na(Region)) %>%
    aggregate(groupby = 'Region',
              output = 'matrix') %>%
    t()
  
  human <- human %>%
    label_human_data(tree = tree_human_pruned,
                     samples = sample_ids) %>%
    aggregate(groupby = 'Region',
              output = 'matrix') %>%
    t()
  
  similarity <- buildSimilarityMatrix(x1 = human,
                                      x2 = mouse,
                                      method = 'correlation')
  
  similarity <- similarity[,match(mouse_regions, colnames(similarity))]
  similarity <- similarity[match(human_regions, rownames(similarity)),]
  
  list_sim[[i]] <- similarity
}

array_sim <- array(unlist(list_sim),
                   dim = c(nrow(similarity),
                           ncol(similarity),
                           length(list_sim)))

mat_sim_avg <- rowMeans(array_sim, dims = 2)
rownames(mat_sim_avg) <- rownames(similarity)
colnames(mat_sim_avg) <- colnames(similarity)
```

```{r fig.width=14}
pheatmap(mat_sim_avg,
         cluster_cols = F,
         cluster_rows = F,
         na_col = 'black')
```

This matrix looks the way I expect it to. This tells me that the latent spaces were generated as intended, and also that transforming regionally aggregated expression data gives the same result as transforming voxel data and then aggregating it. Again, one important difference here is that we've transformed data belonging to the white matter and ventricles. Since the network was only trained to classify grey matter regions, it's possible that voxels belonging to these omitted regions do not transform appropriately. It's hard to say from these regional similarity matrices though. 

# Conclusions

Other than the white matter and ventricular voxels, I'm confident that the input gene space and the latent spaces are behaving as I intend them to. The question then is why are the cluster signatures behaving so differently? In the input space and latent space similarity matrices, we see high correlation within broadly similar regions, but low correlations between those regions. However in the case of the clusters, we find high correlations for all clusters when using the latent space. Why is this happening? Did I make a mistake in the computation of the cluster signatures? 

# Next steps

I want to determine whether I made a mistake when computing the cluster signatures for the different mouse and human clusters. How can I go about doing this? 

I guess the first thing to do might be to dig into the details of the cluster signatures for one or two pairs of clusters at a given threshold. What do the cluster signatures look like? Where are they coming from? Are the human microarray samples being assigned to the correct voxels in the brain mask? 
