---
title: "Human cluster pathway enrichment"
author: "Antoine Beauchamp"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

```{r packages}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(patchwork))
```

```{r environment}
PROJECTPATH <- Sys.getenv("PROJECTPATH")
SRCPATH <- Sys.getenv("SRCPATH")
setwd(PROJECTPATH)
```

```{r functions}
source(file.path(SRCPATH, "analysis.R"))

importMsigDBGMT <- function(file) {
  # stop("This does not work at the present.")
  msig <- list()
  con <- file(file, open = "r")
  lines <- readLines(con)
  close(con)
  ids <- gsub("\t.*", "", lines)
  desc <- gsub("^[^\t]*\t([^\t]*)\t.*", "\\1", lines)
  genes <- gsub("^[^\t]*\t[^\t]*\t(.*)", "\\1", lines)
  msig$MODULES <- data.frame(ID = ids, Title = desc, stringsAsFactors = FALSE)
  if (any(duplicated(msig$MODULES$ID))) {
    warning("Duplicated IDs found; automatic IDs will be generated")
    msig$MODULES$oldID <- msig$MODULES$ID
    msig$MODULES$ID <- make.unique(as.character(msig$MODULES$ID))
  }
  rownames(msig$MODULES) <- msig$MODULES[, "ID"]
  msig$MODULES2GENES <- strsplit(genes, "\t")
  names(msig$MODULES2GENES) <- ids
  msig$GENES <- data.frame(ID = unique(unlist(msig$MODULES2GENES)))
  # msig <- new("tmod", msig)
  msig
}
```

```{r directories}
output_dir <- file.path(PROJECTPATH, "analyses", "primary", "outputs", "human_cluster_pathways")
if (!dir.exists(output_dir)){
  dir.create(output_dir, recursive = TRUE)
}
```


# Pathway module characteristics

```{r pathways-import}
bader_modules <- file.path(PROJECTPATH, "data/human/enrichment/", "Human_Reactome_October_01_2023_symbol.gmt")  
bader_modules <- importMsigDBGMT(bader_modules)

df_modules_size <- map_dbl(bader_modules[["MODULES2GENES"]], length) %>% 
  enframe(name = "ID", value = "B") %>% 
  inner_join(bader_modules[["MODULES"]], by = "ID")
```

```{r pathways-ecdf}
# Sorted pathway sizes
B <- sort(df_modules_size$B)

# Pathway size extrema
B_min <- 0
B_max <- max(B)

# Compute pathway size ECDF
B_seq <- B_min:B_max
B_n <- length(B_seq)
B_ecdf <- numeric(B_n)
for (i in 1:B_n) {B_ecdf[i] <- sum(B <= B_seq[i])/length(B)}

# Indices for size of 10, size at 50%, and size at 90%
idx_0.5 <- min(which(B_ecdf >= 0.5))
idx_0.9 <- min(which(B_ecdf >= 0.9))
idx_10 <- min(which(B_seq >= 10))

# Plot annotations
annotation_x <- B_seq[c(idx_10, idx_0.5, idx_0.9)] + 160
annotation_y <- B_ecdf[c(idx_10, idx_0.5, idx_0.9)]
annotation_labels <- c(paste0("(", B_seq[idx_10], ", ", round(B_ecdf[idx_10],2), ")"),
                       paste0("(", B_seq[idx_0.5], ", ", round(B_ecdf[idx_0.5],2), ")"),
                       paste0("(", B_seq[idx_0.9], ", ", round(B_ecdf[idx_0.9],2), ")"))

# Generate plot
pathways_ecdf <- tibble(B = B_seq,
                        cdf = B_ecdf) %>% 
  ggplot(aes(x = B, y = cdf)) + 
  geom_line() +
  annotate(geom = "point",
           x = B_seq[c(idx_10, idx_0.5, idx_0.9)], 
           y = B_ecdf[c(idx_10, idx_0.5, idx_0.9)]) + 
  annotate(geom = "text",
           label = annotation_labels,
           x = annotation_x, 
           y = annotation_y,
           size = 3.5) + 
  scale_x_continuous(breaks = seq(0, 3000, by = 200)) + 
  scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
  labs(x = "Number of genes in pathway module",
       y = "ECDF",
       title = "Cumulative distribution of pathway sizes") + 
  theme_bw()

# Export plot
outfile <- "pathways_ecdf.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"), 
    height = unit(4, "in"))
print(pathways_ecdf)
dev.off()

# Print plot
pathways_ecdf
```

There are a lot of small pathways.

```{r pathways-small}
n_lt_100 <- df_modules_size %>% 
  filter(B < 100) %>% 
  nrow()

n_lt_10 <- df_modules_size %>% 
  filter(B < 10) %>% 
  nrow()

print(n_lt_100/nrow(df_modules_size))
print(n_lt_10/nrow(df_modules_size))
```

90% of pathways have fewer than 100 genes. 
39% of pathways have fewer than ~10 genes. 

```{r pathways-largest}
df_modules_size %>% 
  arrange(desc(B)) %>% 
  select(Title, B) %>% 
  head(n = 20)
```

```{r pathways-smallest}
df_modules_size %>% 
  arrange(B) %>% 
  select(Title, B)
```

```{r pathways-gt1000}
df_modules_size %>% 
  filter(B >= 1000) %>% 
  nrow()
```

```{r pathways-eq1}
df_modules_size %>% 
  filter(B == 1) %>% 
  nrow()
```

```{r pathways-subset}
# Pathway size threshold
B_threshold <- 10

# Pathways to keep
pathway_ids_keep <- df_modules_size %>% 
  filter(B >= B_threshold) %>% 
  pull(ID)
```


# Cluster pathway enrichment

```{r params}
human_params_id <- 700
human_pipeline_dir <- "data/human/derivatives/v3/"
human_pipeline_dir <- file.path(PROJECTPATH, human_pipeline_dir)

mouse_params_id <- 107
mouse_pipeline_dir <- "data/mouse/derivatives/v3/"
mouse_pipeline_dir <- file.path(PROJECTPATH, mouse_pipeline_dir)

# Gene score for StringDB
gene_score <- 950
```

```{r}
# Import cluster similarity
similarity <- import_similarity(param_id = 375, 
                                pipeline_dir = file.path(PROJECTPATH, "data/cross_species/v3"))

# Import cluster similarity permutations
permutations <- import_similarity_permutations(param_id = 375,
                                               pipeline_dir = file.path(PROJECTPATH, "data/cross_species/v3"))

# Compute significance of correlations
df_sim_pvals <- compute_similarity_significance(similarity = similarity,
                                                permutations = permutations)

colnames(df_sim_pvals) <- str_replace(colnames(df_sim_pvals), "img1", "human")
colnames(df_sim_pvals) <- str_replace(colnames(df_sim_pvals), "img2", "mouse")
```



## Mouse pathway enrichment numbers

```{r}
nk_max <- 10
n_enriched <- c()
for (nk in 2:nk_max) {
  for (k in 1:nk) {
    mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                                pipeline_dir = mouse_pipeline_dir, 
                                                nk = nk, k = k)
    
    mouse_enrichment <- mouse_enrichment %>% 
      filter(ID %in% pathway_ids_keep) %>% 
      arrange(ID)
    
    n_enriched <- c(n_enriched, sum(mouse_enrichment$adj.P.Val < 0.05))
    
  }
}

max(n_enriched)
max(n_enriched)/nrow(mouse_enrichment)
```



## Analysis of human 2-cluster pathway enrichment

### Human 2-cluster with pre-selected pathways

```{r pathway-ids-preselected}
pathway_ids <- c("ADHERENS JUNCTIONS INTERACTIONS%REACTOME DATABASE ID RELEASE 38%418990",
                 "AXON GUIDANCE%REACTOME%R-HSA-422475.7",
                 "CA2+ PATHWAY%REACTOME%R-HSA-4086398.4",
                 "CHROMATIN ORGANIZATION%REACTOME DATABASE ID RELEASE 38%4839726",
                 "GAP JUNCTION TRAFFICKING AND REGULATION%REACTOME%R-HSA-157858.2",
                 "GENE EXPRESSION (TRANSCRIPTION)%REACTOME%R-HSA-74160.8",
                 "GENERIC TRANSCRIPTION PATHWAY%REACTOME%R-HSA-212436.12",
                 "LONG-TERM POTENTIATION%REACTOME DATABASE ID RELEASE 38%9620244",
                 "MAPK FAMILY SIGNALING CASCADES%REACTOME%R-HSA-5683057.4",
                 "MTOR SIGNALLING%REACTOME%R-HSA-165159.7",
                 "PROTEIN-PROTEIN INTERACTIONS AT SYNAPSES%REACTOME DATABASE ID RELEASE 38%6794362",
                 "SIGNALING BY ERBB2%REACTOME DATABASE ID RELEASE 38%1227986",
                 "SIGNALING BY ERBB4%REACTOME DATABASE ID RELEASE 38%1236394",
                 "SIGNALING BY GPCR%REACTOME DATABASE ID RELEASE 38%372790",
                 "SIGNALING BY HEDGEHOG%REACTOME DATABASE ID RELEASE 38%5358351",
                 "SIGNALING BY NOTCH%REACTOME%R-HSA-157118.6",
                 "SIGNALING BY VEGF%REACTOME%R-HSA-194138.3",
                 "SIGNALING BY WNT%REACTOME DATABASE ID RELEASE 38%195721",
                 "TIGHT JUNCTION INTERACTIONS%REACTOME DATABASE ID RELEASE 38%420029",
                 "TRANSMISSION ACROSS CHEMICAL SYNAPSES%REACTOME%R-HSA-112315.7")
```

```{r}
nk_human <- 2; k_human <- 1
nk_mouse <- 4; k_mouse <- 1

human_enrichment <- import_enrichment_human(params_id = human_params_id, 
                                            pipeline_dir = human_pipeline_dir, 
                                            nk = nk_human, k = k_human)

mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                            pipeline_dir = mouse_pipeline_dir, 
                                            nk = nk_mouse, k = k_mouse)

human_enrichment <- human_enrichment %>% 
  filter(ID %in% pathway_ids) %>% 
  arrange(ID)

mouse_enrichment <- mouse_enrichment %>% 
  filter(ID %in% pathway_ids) %>% 
  arrange(ID)

mouse_alpha_seq_auc <- 2:10
mouse_alpha_seq_auc <- (1/10)^mouse_alpha_seq_auc
mouse_alpha_seq_auc <- c(0.05, mouse_alpha_seq_auc)

enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                         human_enrichment = human_enrichment,
                                         alpha = mouse_alpha_seq_auc)

enrichment_AUC <- map_dbl(enrichment_ROC, pROC::auc)

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha = mouse_alpha_seq_auc)
```


```{r}
df_roc_all %>% 
  mutate(nlq = -log10(mouse_alpha),
         nlq = factor(nlq)) %>% 
  ggplot(aes(x = fpr, y = tpr, col = nlq)) + 
  annotate(geom = "segment",
           x = 0, xend = 1,
           y = 0, yend = 1,
           size = 0.3,
           linetype = "dashed",
           col = "black") +
  geom_line(size = 0.35) +
  # facet_wrap(~NLQ_mouse_lab, nrow = 2, ncol = 2) + 
  scale_x_continuous(limits = c(0, 1)) + 
  scale_y_continuous(limits = c(0, 1), 
                     expand = expansion(add = 0.1)) + 
  # scale_color_manual(values = pathways_human_palette) + 
  labs(x = "False positive rate",
       y = "True positive rate") +
  theme_bw()
```

```{r}
mouse_alpha_seq_auc <- 2:10
mouse_alpha_seq_auc <- (1/10)^mouse_alpha_seq_auc
mouse_alpha_seq_auc <- c(0.05, mouse_alpha_seq_auc)

df_auc <- tibble(alpha = mouse_alpha_seq_auc,
                 auc = 0)
for (i in 1:nrow(df_auc)) {
  
  predictor <- human_enrichment %>% 
    pull(adj.P.Val)
  
  outcome <- mouse_enrichment %>% 
    mutate(pass = adj.P.Val < df_auc[[i, "alpha"]]) %>% 
    pull(pass)
  
  # Compute AUC
  df_auc[[i, "auc"]] <- pROC::auc(outcome, 1-predictor, quiet = TRUE)[1]
  
}

ggplot(df_auc, aes(x = -log10(alpha), y = auc)) + 
  geom_line() + 
  geom_point() + 
  coord_cartesian(ylim = c(0.5, 1)) + 
  scale_x_continuous(breaks = seq(2.0, 10.0, by = 1)) + 
  theme_bw()
```


### Human 2-cluster with all pathways

```{r}
human_enrichment <- import_enrichment_human(params_id = human_params_id, 
                                            pipeline_dir = human_pipeline_dir, 
                                            nk = 2, k = 1)

mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                            pipeline_dir = mouse_pipeline_dir, 
                                            nk = 4, k = 1)

human_enrichment <- human_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)

mouse_enrichment <- mouse_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)

mouse_alpha_seq <- 2:10
mouse_alpha_seq <- (1/10)^mouse_alpha_seq
mouse_alpha_seq <- c(0.05, mouse_alpha_seq)

enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                         human_enrichment = human_enrichment,
                                         alpha = mouse_alpha_seq)

enrichment_ROC_AUC <- map_dbl(enrichment_ROC, pROC::auc) %>% 
  enframe(name = "alpha", value = "AUC") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                       human_enrichment = human_enrichment,
                                       alpha = mouse_alpha_seq)

enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                             .f = function(x){
                               tibble(AUC = x[["auc.integral"]], 
                                      Prop = x[["true.proportion"]])
                             }, .id = "alpha") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha = mouse_alpha_seq)

head(enrichment_HG)
```

```{r}
plt_nk2_ROC_AUC <- ggplot(enrichment_ROC_AUC, 
                          aes(x = -log10(alpha), y = AUC)) +
  geom_line() + 
  geom_hline(yintercept = 0.5, linetype = "dashed") + 
  geom_point() + 
  coord_cartesian(ylim = c(0.5, 1.0)) + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) +
  labs(x = "Mouse threshold (-log10(q))",
       y = "ROC AUC",
       title = "ROC pathway comparison between human 2-1 and mouse 4-1") + 
  theme_bw()

outfile <- "pathway_comparison_nk2_ROC_AUC.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(6, "in"),
    height = unit(3, "in"))
print(plt_nk2_ROC_AUC)
dev.off()

plt_nk2_ROC_AUC
```

```{r}
plt_nk2_PR_AUC <- enrichment_PR_AUC %>% 
  rename(Chance = Prop) %>% 
  pivot_longer(cols = -alpha, names_to = "measure", values_to = "values") %>% 
  ggplot(aes(x = -log10(alpha), y = values)) +
  geom_line(aes(group = measure, linetype = measure)) + 
  geom_point() + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  labs(x = "Mouse threshold (-log10(q))",
       y = "PR AUC",
       title = "Precision-recall pathway comparison between human 2-1 and mouse 4-1") + 
  theme_bw() +
  theme(title = element_text(size = 7))

outfile <- "pathway_comparison_nk2_PR_AUC.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(6, "in"),
    height = unit(3, "in"))
print(plt_nk2_PR_AUC)
dev.off()

plt_nk2_PR_AUC
```

```{r}
plt_nk2_HG <- enrichment_HG %>% 
  ggplot(aes(x = -log10(alpha_human), y = -log10(pval), col = factor(round(-log10(alpha_mouse), 2)))) + 
  geom_line() + 
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  labs(x = "Human threshold (-log10(q))",
       y = "-log10(p)",
       col = "Mouse threshold (-log10(q))",
       title = "Hypergeometric test pathway comparison between human 2-1 and mouse 4-1") + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  scale_y_continuous(breaks = seq(0, 20, by = 2)) + 
  theme_bw()

outfile <- "pathway_comparison_nk2_HG.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"),
    height = unit(4, "in"))
print(plt_nk2_HG)
dev.off()

plt_nk2_HG
```
#### Random chance

```{r}
sum(mouse_enrichment$adj.P.Val < 0.05)
sum(human_enrichment$adj.P.Val < 0.05)
```

```{r}
n_sim <- 500
mouse_alpha_seq <- c(0.05, 1e-2, 1e-4, 1e-6, 1e-8, 1e-10)
list_ROC_AUC <- vector(mode = "list", length = length(mouse_alpha_seq))
list_PR_AUC <- vector(mode = "list", length = length(mouse_alpha_seq))
list_HG <- vector(mode = "list", length = length(mouse_alpha_seq))
for (i in 1:length(mouse_alpha_seq)) {
  
  alpha <- mouse_alpha_seq[i]
  n_true <- sum(mouse_enrichment$adj.P.Val < alpha)
  
  list_ROC_AUC_i <- vector(mode = "list", length = n_sim)
  list_PR_AUC_i <- vector(mode = "list", length = n_sim)
  list_HG_i <- vector(mode = "list", length = n_sim)
  for (s in 1:n_sim) {
    
    mouse_enrichment_test <- mouse_enrichment %>% mutate(adj.P.Val = 1)
    set.seed(s)
    idx_flip <- sample(1:nrow(mouse_enrichment_test), size = n_true, replace = FALSE)
    mouse_enrichment_test$adj.P.Val[idx_flip] <- 0
    
    
    enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment_test, 
                                             human_enrichment = human_enrichment,
                                             alpha = alpha)
    
    enrichment_ROC_AUC <- map_dbl(enrichment_ROC, pROC::auc) %>% 
      enframe(name = "alpha", value = "AUC") %>% 
      mutate(alpha = as.numeric(alpha))
    
    list_ROC_AUC_i[[s]] <- enrichment_ROC_AUC
    
    enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment_test,
                                           human_enrichment = human_enrichment,
                                           alpha = alpha)
    
    enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                                 .f = function(x){
                                   tibble(AUC = x[["auc.integral"]], 
                                          Prop = x[["true.proportion"]])
                                 }, .id = "alpha") %>% 
      mutate(alpha = as.numeric(alpha))
    
    list_PR_AUC_i[[s]] <- enrichment_PR_AUC
    
    enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment_test, 
                                               human_enrichment = human_enrichment,
                                               alpha_mouse = alpha,
                                               alpha_human = c(0.05, 1e-3, 1e-5, 1e-10))
    
    list_HG_i[[s]] <- enrichment_HG
    
  }
  
  list_ROC_AUC[[i]] <- bind_rows(list_ROC_AUC_i, .id = "sample")
  list_PR_AUC[[i]] <- bind_rows(list_PR_AUC_i, .id = "sample")
  list_HG[[i]] <- bind_rows(list_HG_i, .id = "sample")
  
}

df_ROC_AUC <- bind_rows(list_ROC_AUC)
df_PR_AUC <- bind_rows(list_PR_AUC)
df_HG <- bind_rows(list_HG)
```


```{r}
ggplot(df_ROC_AUC, aes(x = -log10(alpha), y = AUC)) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.05) + 
  geom_hline(yintercept = 0.5, linetype = "dashed", col = "red", linewidth = 1) + 
  coord_cartesian(ylim = c(0.3, 1)) + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  labs(x = "Mouse threshold (-log10(q))",
       y = "ROC AUC") + 
  theme_bw()
```

```{r}
df_PR_true_prop <- df_PR_AUC %>% select(alpha, Prop) %>% distinct()

ggplot(df_PR_AUC, aes(x = -log10(alpha), y = AUC)) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.1) + 
  geom_line(data = df_PR_true_prop, 
            mapping = aes(x = -log10(alpha), y = Prop),
            col = "red", linetype = "dashed", linewidth = 1) + 
  # coord_cartesian(ylim = c(0.3, 1)) + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  labs(x = "Mouse threshold (-log10(q))",
       y = "PR AUC") + 
  theme_bw()
```

```{r}
tmp <- df_HG %>% 
  filter(alpha_human == 0.05)

ggplot(df_HG, aes(x = -log10(alpha_mouse), y = -log10(pval))) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.5) + 
  geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed") + 
  facet_wrap(~round(-log10(alpha_human), 2)) +
  # facet_wrap(~alpha_human) + 
  # coord_cartesian(ylim = c(0.3, 1)) + 
  scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  labs(x = "Mouse threshold (-log10(q))",
       y = "-log10(p)") + 
  theme_bw()
```


```{r}
enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                       human_enrichment = human_enrichment,
                                       alpha = mouse_alpha)

enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                             .f = function(x){
                               tibble(AUC = x[["auc.integral"]], 
                                      Prop = x[["true.proportion"]])
                             }, .id = "alpha") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha = mouse_alpha_seq)
```


## Comparison between human 7-2 and mouse nk = 7

```{r}
# Human cluster to examine
human_cluster_id <- "7-2"
human_nk <- str_split(human_cluster_id, "-")[[1]][1]
human_k <- str_split(human_cluster_id, "-")[[1]][2]

# Import human pathway enrichment for current cluster
human_enrichment <- import_enrichment_human(params_id = human_params_id, 
                                            pipeline_dir = human_pipeline_dir, 
                                            nk = human_nk, k = human_k)

# Filter for pathways to keep and arrange by ID for correspondence to mouse
human_enrichment <- human_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)
```


### Random chance behaviour for human 7-2

```{r}
# Sequence of number of pathways enriched in cluster
n_enriched_seq <- c(300, 100, 50, 25, 10, 5:1)

df_ROC_AUC <- simulate_enrichment_null(enrichment = human_enrichment,
                                       n_true = n_enriched_seq,
                                       method = "ROC")

df_PR_AUC <- simulate_enrichment_null(enrichment = human_enrichment,
                                      n_true = n_enriched_seq,
                                      method = "PR")

df_HG <- simulate_enrichment_null(enrichment = human_enrichment,
                                  n_true = n_enriched_seq,
                                  method = "HG")

df_ROC_AUC_quantiles <- df_ROC_AUC %>% 
  group_by(n_enriched) %>% 
  summarise(AUC_lwr = quantile(AUC, probs = 0.025),
            AUC_upr = quantile(AUC, probs = 0.975))

df_PR_AUC_quantiles <- df_PR_AUC %>% 
  group_by(n_enriched) %>% 
  summarise(AUC_chance = mean(Prop),
            AUC_lwr = quantile(AUC, probs = 0.025),
            AUC_upr = quantile(AUC, probs = 0.975))
```

```{r}
plt_H72_ROC_AUC_chance <- ggplot(df_ROC_AUC, aes(x = n_enriched, y = AUC)) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.05) + 
  geom_hline(yintercept = 0.5, linetype = "dashed", col = "red", linewidth = 1) + 
  coord_cartesian(ylim = c(0.4, 1)) + 
  scale_x_continuous(breaks = seq(0, 300, by = 20)) + 
  scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
  labs(x = "Number of mouse pathways enriched",
       y = "ROC AUC",
       title = "Random chance behaviour of ROC AUC for human cluster 7-2") + 
  theme_bw()

# Export plot
outfile <- "human_7-2_ROC_AUC_random_chance.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"), 
    height = unit(4, "in"))
print(plt_H72_ROC_AUC_chance)
dev.off()

plt_H72_ROC_AUC_chance
```

```{r}
plt_H72_PR_AUC_chance <- ggplot(df_PR_AUC, aes(x = n_enriched, y = AUC)) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.1) + 
  geom_abline(intercept = 0, slope = 1/nrow(human_enrichment),
              col = "red", linetype = "dashed", linewidth = 1) +
  scale_x_continuous(breaks = seq(0, 500, by = 20)) + 
  labs(x = "Number of mouse pathways enriched",
       y = "PR AUC",
       title = "Random chance behaviour of PR AUC for human cluster 7-2") + 
  theme_bw()

# Export plot
outfile <- "human_7-2_PR_AUC_random_chance.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"), 
    height = unit(4, "in"))
print(plt_H72_PR_AUC_chance)
dev.off()

plt_H72_PR_AUC_chance
```
```{r}
ggplot(df_HG, aes(x = B, y = -log10(pval))) + 
  geom_line(aes(group = sample),
            alpha = 0.03) + 
  geom_point(alpha = 0.5) +
  geom_hline(yintercept = -log10(0.05), col = "red", linetype = "dashed") + 
  facet_wrap(~round(-log10(alpha_human), 2)) +
  # facet_wrap(~alpha_human) + 
  # coord_cartesian(ylim = c(0.3, 1)) + 
  # scale_x_continuous(breaks = round(-log10(mouse_alpha_seq), 2)) + 
  labs(x = "Mouse threshold (-log10(q))",
       y = "-log10(p)") + 
  theme_bw()
```

```{r}
df_HG_sig_prop <- df_HG %>% 
  mutate(significant = pval < 0.05) %>% 
  group_by(alpha_human, B) %>% 
  summarise(n_sig = sum(significant),
            prop_sig = n_sig/n_sim, 
            .groups = "drop")

ggplot(df_HG_sig_prop, aes(x = B, y = prop_sig, col = factor(-log10(alpha_human)))) + 
  geom_point()
```


### Best cluster match: Human 7-2 to mouse 7-1

```{r}
human_clust_id <- human_cluster_id
df_sim_pvals %>% 
  filter(human_cluster_id == human_clust_id, mouse_nk == 7) %>% 
  select(mouse_cluster_id, similarity, pval) %>% 
  arrange(pval)
```

```{r}
mouse_cluster_id <- "7-1"
mouse_nk <- str_split(mouse_cluster_id, "-")[[1]][1]
mouse_k <- str_split(mouse_cluster_id, "-")[[1]][2]

mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                            pipeline_dir = mouse_pipeline_dir, 
                                            nk = mouse_nk, k = mouse_k)

mouse_enrichment <- mouse_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)

mouse_alpha_seq <- 2:10
mouse_alpha_seq <- (1/10)^mouse_alpha_seq
mouse_alpha_seq <- c(0.05, mouse_alpha_seq)

enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                         human_enrichment = human_enrichment,
                                         alpha = mouse_alpha_seq)

enrichment_ROC_AUC <- inner_join(map_dbl(enrichment_ROC, pROC::auc) %>% 
                                   enframe(name = "alpha", value = "AUC") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 map_dbl(enrichment_ROC, function(x){sum(x[["original.response"]])}) %>% 
                                   enframe(name = "alpha", value = "n_enriched") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 by = "alpha")

enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                       human_enrichment = human_enrichment,
                                       alpha = mouse_alpha_seq)

enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                             .f = function(x){
                               tibble(AUC = x[["auc.integral"]], 
                                      prop = x[["true.proportion"]],
                                      n_enriched = prop*nrow(human_enrichment))
                             }, .id = "alpha") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha_mouse = mouse_alpha_seq)
```

```{r}
enrichment_ROC_AUC <- enrichment_ROC_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M71_ROC_AUC <- ggplot(enrichment_ROC_AUC, 
                              aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() + 
  geom_ribbon(data = df_ROC_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_hline(yintercept = 0.5, linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, 130),
                  ylim = c(0.4, 1.0)) +
  scale_x_continuous(breaks = seq(0, 200, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_ROC_AUC$n_enriched,
                                         labels = enrichment_ROC_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) + 
  scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
  labs(x = "Number of mouse pathways enriched",
       y = "ROC AUC",
       title = "ROC pathway comparison between human 7-2 and mouse 7-1") + 
  theme_bw()

# Export plot
# outfile <- "human_7-2_mouse_7-1_ROC_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"), 
#     height = unit(4, "in"))
# print(plt_H72_M71_ROC_AUC)
# dev.off()

plt_H72_M71_ROC_AUC
```

```{r}
enrichment_PR_AUC <- enrichment_PR_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M71_PR_AUC <- ggplot(enrichment_PR_AUC, 
                             aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() +
  geom_ribbon(data = df_PR_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_line(data = df_PR_AUC_quantiles, 
            mapping = aes(y = AUC_chance),
            linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, 130)) +
  scale_x_continuous(breaks = seq(0, 200, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  labs(x = "Number of mouse pathways enriched",
       y = "PR AUC",
       title = "Precision-recall pathway comparison between human 7-2 and mouse 7-1") + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-1_PR_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M71_PR_AUC)
# dev.off()

plt_H72_M71_PR_AUC
```

```{r}
palette <- colorRampPalette(rev(brewer.pal(n = 9, name = "Reds")[3:9]))(10)

plt_H72_M71_HG <- enrichment_HG %>% 
  ggplot(aes(x = B, y = -log10(pval), col = factor(round(-log10(alpha_human), 2)))) + 
  geom_line() + 
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  labs(col = "Human threshold\n(-log10(q))",
       y = "-log10(p)",
       x = "Number of mouse pathways enriched",
       title = "Hypergeometric test pathway comparison between human 7-2 and mouse 7-1") + 
  coord_cartesian(xlim = c(0, 130)) +
  scale_x_continuous(breaks = seq(0, 200, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  scale_y_continuous(breaks = seq(0, 20, by = 1)) + 
  scale_color_manual(values = palette) + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-1_HG.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M71_HG)
# dev.off()

plt_H72_M71_HG
```

```{r}
cluster_r <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(similarity) %>% 
  round(3)

cluster_p <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(pval) %>% 
  round(3)

cluster_stats <- paste0("Transcriptomic similarity: ", "r = ", cluster_r, "; p = ", cluster_p)

plt_H72_M71_ROC_AUC_pw <- plt_H72_M71_ROC_AUC +
  labs(title = "ROC comparison")

plt_H72_M71_PR_AUC_pw <- plt_H72_M71_PR_AUC +
  labs(title = "Precision-recall comparison") 

plt_H72_M71_HG_pw <- plt_H72_M71_HG +
  labs(title = "Hypergeometric test comparison") + 
  theme(legend.position = "bottom")

plt_H72_M71 <- plt_H72_M71_ROC_AUC_pw / plt_H72_M71_PR_AUC_pw / plt_H72_M71_HG_pw +
  plot_annotation(title = "Pathway correspondence between human 7-2 and mouse 7-1",
                  subtitle = cluster_stats)

outfile <- "human_7-2_mouse_7-1.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"),
    height = unit(10, "in"))
print(plt_H72_M71)
dev.off()

plt_H72_M71
```


### Worst cluster match: Human 7-2 to mouse 7-7

```{r}
mouse_cluster_id <- "7-7"
mouse_nk <- str_split(mouse_cluster_id, "-")[[1]][1]
mouse_k <- str_split(mouse_cluster_id, "-")[[1]][2]

mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                            pipeline_dir = mouse_pipeline_dir, 
                                            nk = mouse_nk, k = mouse_k)

mouse_enrichment <- mouse_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)

mouse_alpha_seq <- 2:10
mouse_alpha_seq <- (1/10)^mouse_alpha_seq
mouse_alpha_seq <- c(0.05, mouse_alpha_seq)

enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                         human_enrichment = human_enrichment,
                                         alpha = mouse_alpha_seq)

enrichment_ROC_AUC <- inner_join(map_dbl(enrichment_ROC, pROC::auc) %>% 
                                   enframe(name = "alpha", value = "AUC") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 map_dbl(enrichment_ROC, function(x){sum(x[["original.response"]])}) %>% 
                                   enframe(name = "alpha", value = "n_enriched") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 by = "alpha")

enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                       human_enrichment = human_enrichment,
                                       alpha = mouse_alpha_seq)

enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                             .f = function(x){
                               tibble(AUC = x[["auc.integral"]], 
                                      prop = x[["true.proportion"]],
                                      n_enriched = prop*nrow(human_enrichment))
                             }, .id = "alpha") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha_mouse = mouse_alpha_seq)

xlim_max <- round(max(enrichment_ROC_AUC$n_enriched), -1) + 10
```

```{r}
enrichment_ROC_AUC <- enrichment_ROC_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M77_ROC_AUC <- ggplot(enrichment_ROC_AUC, 
                              aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() + 
  geom_ribbon(data = df_ROC_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_hline(yintercept = 0.5, linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, xlim_max),
                  ylim = c(0.4, 1.0)) +
  scale_x_continuous(breaks = seq(0, 200, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_ROC_AUC$n_enriched,
                                         labels = enrichment_ROC_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) + 
  scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
  labs(x = "Number of mouse pathways enriched",
       y = "ROC AUC",
       title = "ROC pathway comparison between human 7-2 and mouse 7-7") + 
  theme_bw()

# Export plot
# outfile <- "human_7-2_mouse_7-7_ROC_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"), 
#     height = unit(4, "in"))
# print(plt_H72_M77_ROC_AUC)
# dev.off()

plt_H72_M77_ROC_AUC
```

```{r}
enrichment_PR_AUC <- enrichment_PR_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M77_PR_AUC <- ggplot(enrichment_PR_AUC, 
                             aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() +
  geom_ribbon(data = df_PR_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_line(data = df_PR_AUC_quantiles, 
            mapping = aes(y = AUC_chance),
            linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, xlim_max)) +
  scale_x_continuous(breaks = seq(0, 300, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  labs(x = "Number of mouse pathways enriched",
       y = "PR AUC",
       title = "Precision-recall pathway comparison between human 7-2 and mouse 7-7") + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-7_PR_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M77_PR_AUC)
# dev.off()

plt_H72_M77_PR_AUC
```

```{r}
palette <- colorRampPalette(rev(brewer.pal(n = 9, name = "Reds")[3:9]))(10)

plt_H72_M77_HG <- enrichment_HG %>% 
  ggplot(aes(x = B, y = -log10(pval), col = factor(round(-log10(alpha_human), 2)))) + 
  geom_line() + 
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  labs(col = "Human threshold\n(-log10(q))",
       y = "-log10(p)",
       x = "Number of mouse pathways enriched",
       title = "Hypergeometric test pathway comparison between human 7-2 and mouse 7-7") + 
  coord_cartesian(xlim = c(0, xlim_max)) +
  scale_x_continuous(breaks = seq(0, 300, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  scale_y_continuous(breaks = seq(0, 20, by = 1)) + 
  scale_color_manual(values = palette) + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-7_HG.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M77_HG)
# dev.off()

plt_H72_M77_HG
```

```{r}
cluster_r <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(similarity) %>% 
  round(3)

cluster_p <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(pval) %>% 
  round(3)

cluster_stats <- paste0("Transcriptomic similarity: ", "r = ", cluster_r, "; p = ", cluster_p)

plt_H72_M77_ROC_AUC_pw <- plt_H72_M77_ROC_AUC +
  labs(title = "ROC comparison")

plt_H72_M77_PR_AUC_pw <- plt_H72_M77_PR_AUC +
  labs(title = "Precision-recall comparison") 

plt_H72_M77_HG_pw <- plt_H72_M77_HG +
  labs(title = "Hypergeometric test comparison") + 
  theme(legend.position = "bottom")

plt_H72_M77 <- plt_H72_M77_ROC_AUC_pw / plt_H72_M77_PR_AUC_pw / plt_H72_M77_HG_pw +
  plot_annotation(title = "Pathway correspondence between human 7-2 and mouse 7-7",
                  subtitle = cluster_stats)

outfile <- "human_7-2_mouse_7-7.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"),
    height = unit(10, "in"))
print(plt_H72_M77)
dev.off()

plt_H72_M77
```


### Human 7-2 to mouse 7-4

```{r}
mouse_cluster_id <- "7-4"
mouse_nk <- str_split(mouse_cluster_id, "-")[[1]][1]
mouse_k <- str_split(mouse_cluster_id, "-")[[1]][2]

mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                            pipeline_dir = mouse_pipeline_dir, 
                                            nk = mouse_nk, k = mouse_k)

mouse_enrichment <- mouse_enrichment %>% 
  filter(ID %in% pathway_ids_keep) %>% 
  arrange(ID)

mouse_alpha_seq <- 2:10
mouse_alpha_seq <- (1/10)^mouse_alpha_seq
mouse_alpha_seq <- c(0.05, mouse_alpha_seq)

enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                         human_enrichment = human_enrichment,
                                         alpha = mouse_alpha_seq)

enrichment_ROC_AUC <- inner_join(map_dbl(enrichment_ROC, pROC::auc) %>% 
                                   enframe(name = "alpha", value = "AUC") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 map_dbl(enrichment_ROC, function(x){sum(x[["original.response"]])}) %>% 
                                   enframe(name = "alpha", value = "n_enriched") %>% 
                                   mutate(alpha = as.numeric(alpha)),
                                 by = "alpha")

enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                       human_enrichment = human_enrichment,
                                       alpha = mouse_alpha_seq)

enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                             .f = function(x){
                               tibble(AUC = x[["auc.integral"]], 
                                      prop = x[["true.proportion"]],
                                      n_enriched = prop*nrow(human_enrichment))
                             }, .id = "alpha") %>% 
  mutate(alpha = as.numeric(alpha))

enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                           human_enrichment = human_enrichment,
                                           alpha_mouse = mouse_alpha_seq)

xlim_max <- round(max(enrichment_ROC_AUC$n_enriched), -1) + 10
```

```{r}
enrichment_ROC_AUC <- enrichment_ROC_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M74_ROC_AUC <- ggplot(enrichment_ROC_AUC, 
                              aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() + 
  geom_ribbon(data = df_ROC_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_hline(yintercept = 0.5, linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, xlim_max),
                  ylim = c(0.4, 1.0)) +
  scale_x_continuous(breaks = seq(0, 200, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_ROC_AUC$n_enriched,
                                         labels = enrichment_ROC_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) + 
  scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
  labs(x = "Number of mouse pathways enriched",
       y = "ROC AUC",
       title = "ROC pathway comparison between human 7-2 and mouse 7-4") + 
  theme_bw()

# Export plot
# outfile <- "human_7-2_mouse_7-7_ROC_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"), 
#     height = unit(4, "in"))
# print(plt_H72_M74_ROC_AUC)
# dev.off()

plt_H72_M74_ROC_AUC
```

```{r}
enrichment_PR_AUC <- enrichment_PR_AUC %>% 
  mutate(NL = round(-log10(alpha), 2))

plt_H72_M74_PR_AUC <- ggplot(enrichment_PR_AUC, 
                             aes(x = n_enriched, y = AUC)) + 
  geom_line() + 
  geom_point() +
  geom_ribbon(data = df_PR_AUC_quantiles, 
              mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
              inherit.aes = FALSE,
              alpha = 0.3, fill = "red") + 
  geom_line(data = df_PR_AUC_quantiles, 
            mapping = aes(y = AUC_chance),
            linetype = "dashed", col = "red") + 
  coord_cartesian(xlim = c(0, xlim_max)) +
  scale_x_continuous(breaks = seq(0, 300, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  labs(x = "Number of mouse pathways enriched",
       y = "PR AUC",
       title = "Precision-recall pathway comparison between human 7-2 and mouse 7-4") + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-7_PR_AUC.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M74_PR_AUC)
# dev.off()

plt_H72_M74_PR_AUC
```

```{r}
palette <- colorRampPalette(rev(brewer.pal(n = 9, name = "Reds")[3:9]))(10)

plt_H72_M74_HG <- enrichment_HG %>% 
  ggplot(aes(x = B, y = -log10(pval), col = factor(round(-log10(alpha_human), 2)))) + 
  geom_line() + 
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  labs(col = "Human threshold\n(-log10(q))",
       y = "-log10(p)",
       x = "Number of mouse pathways enriched",
       title = "Hypergeometric test pathway comparison between human 7-2 and mouse 7-4") + 
  coord_cartesian(xlim = c(0, xlim_max)) +
  scale_x_continuous(breaks = seq(0, 300, by = 20),
                     sec.axis = dup_axis(breaks = enrichment_PR_AUC$n_enriched,
                                         labels = enrichment_PR_AUC$NL,
                                         name = "Mouse threshold (-log10(q))")) +
  scale_y_continuous(breaks = seq(0, 20, by = 1)) + 
  scale_color_manual(values = palette) + 
  theme_bw()

# outfile <- "human_7-2_mouse_7-7_HG.pdf"
# outfile <- file.path(output_dir, outfile)
# pdf(file = outfile,
#     width = unit(8, "in"),
#     height = unit(4, "in"))
# print(plt_H72_M74_HG)
# dev.off()

plt_H72_M74_HG
```

```{r}
cluster_r <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(similarity) %>% 
  round(3)

cluster_p <- df_sim_pvals %>% 
  filter(human_cluster_id == human_cluster_id, 
         mouse_cluster_id == mouse_cluster_id) %>% 
  pull(pval) %>% 
  round(3)

cluster_stats <- paste0("Transcriptomic similarity: ", "r = ", cluster_r, "; p = ", cluster_p)

plt_H72_M74_ROC_AUC_pw <- plt_H72_M74_ROC_AUC +
  labs(title = "ROC comparison")

plt_H72_M74_PR_AUC_pw <- plt_H72_M74_PR_AUC +
  labs(title = "Precision-recall comparison") 

plt_H72_M74_HG_pw <- plt_H72_M74_HG +
  labs(title = "Hypergeometric test comparison") + 
  theme(legend.position = "bottom")

plt_H72_M74 <- plt_H72_M74_ROC_AUC_pw / plt_H72_M74_PR_AUC_pw / plt_H72_M74_HG_pw +
  plot_annotation(title = "Pathway correspondence between human 7-2 and mouse 7-4",
                  subtitle = cluster_stats)

outfile <- "human_7-2_mouse_7-4.pdf"
outfile <- file.path(output_dir, outfile)
pdf(file = outfile,
    width = unit(8, "in"),
    height = unit(10, "in"))
print(plt_H72_M74)
dev.off()

plt_H72_M74
```
## All clusters?

```{r}
df_matching_clusters <- df_sim_pvals %>% 
  filter(pval < 0.05)

```


```{r}
human_variants <- file.path(human_pipeline_dir, human_params_id, "cluster_variants.csv")
df_human_variants <- read_csv(human_variants, show_col_types = FALSE)

df_human_cluster_ngenes <- df_human_variants %>% 
  select(Subject_ID, file, contains("nk"), gene) %>% 
  filter(!is.na(gene)) %>% 
  pivot_longer(cols = contains("nk"), names_to = "nk", values_to = "k") %>% 
  mutate(nk = as.numeric(str_remove(nk, "nk"))) %>% 
  group_by(nk, k) %>% 
  count() %>% 
  unite(col = "cluster_id", nk, k, sep = "-", remove = FALSE) %>% 
  arrange(nk, k)


df_matching_clusters_variants <- inner_join(df_matching_clusters,
                                            df_human_cluster_ngenes,
                                            by = c("human_cluster_id" = "cluster_id",
                                                   "human_nk" = "nk",
                                                   "human_k" = "k"))


df_matching_clusters_variants <- df_matching_clusters_variants %>% 
  filter(n >= 3, human_nk >= 4)

human_cluster_ids <- df_matching_clusters_variants %>% 
  pull(human_cluster_id) %>% 
  unique()
```


### All clusters plots

```{r}
get_auc_null_quantiles <- function(enrichment, n_true, probs = c(0.025, 0.975)) {
  
  df_ROC_AUC <- simulate_enrichment_null(enrichment = enrichment,
                                         n_true = n_true,
                                         method = "ROC")
  
  df_PR_AUC <- simulate_enrichment_null(enrichment = human_enrichment,
                                        n_true = n_true,
                                        method = "PR")
  
  df_ROC_AUC_quantiles <- df_ROC_AUC %>% 
    group_by(n_enriched) %>% 
    summarise(AUC_lwr = quantile(AUC, probs = probs[1]),
              AUC_upr = quantile(AUC, probs = probs[2]))
  
  df_PR_AUC_quantiles <- df_PR_AUC %>% 
    group_by(n_enriched) %>% 
    summarise(AUC_chance = mean(Prop),
              AUC_lwr = quantile(AUC, probs = probs[1]),
              AUC_upr = quantile(AUC, probs = probs[2]))
  
  return(list(ROC = df_ROC_AUC_quantiles,
              PR = df_PR_AUC_quantiles))
  
}

plot_enrichment_ROC_AUC <- function(x, quantiles, human_cluster, mouse_cluster, export = FALSE) {
  
  x <- mutate(x, NL = round(-log10(alpha), 2))  
  
  xlim_max <- round(max(x$n_enriched), -1) + 10
  
  plt <- ggplot(x, aes(x = n_enriched, y = AUC)) + 
    geom_line() + 
    geom_point() + 
    geom_ribbon(data = quantiles, 
                mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
                inherit.aes = FALSE,
                alpha = 0.3, fill = "red") + 
    geom_hline(yintercept = 0.5, linetype = "dashed", col = "red") + 
    coord_cartesian(xlim = c(0, xlim_max),
                    ylim = c(0.4, 1.0)) +
    scale_x_continuous(breaks = seq(0, 1000, by = 20),
                       sec.axis = dup_axis(breaks = x$n_enriched,
                                           labels = x$NL,
                                           name = "Mouse threshold (-log10(q))")) + 
    scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
    labs(x = "Number of mouse pathways enriched",
         y = "ROC AUC",
         title = paste("ROC pathway comparison between human", human_cluster, "and mouse", mouse_cluster)) + 
    theme_bw()
  
  # Export plot
  # outfile <- "human_7-2_mouse_7-1_ROC_AUC.pdf"
  # outfile <- file.path(output_dir, outfile)
  # pdf(file = outfile,
  #     width = unit(8, "in"), 
  #     height = unit(4, "in"))
  # print(plt_H72_M71_ROC_AUC)
  # dev.off()
  
  return(plt)
  
}

plot_enrichment_PR_AUC <- function(x, quantiles, human_cluster, mouse_cluster, export = FALSE) {
  
  x <- x %>% 
    mutate(NL = round(-log10(alpha), 2))
  
  xlim_max <- round(max(x$n_enriched), -1) + 10
  
  plt <- ggplot(x, aes(x = n_enriched, y = AUC)) + 
    geom_line() + 
    geom_point() +
    geom_ribbon(data = quantiles, 
                mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
                inherit.aes = FALSE,
                alpha = 0.3, fill = "red") + 
    geom_line(data = quantiles, 
              mapping = aes(y = AUC_chance),
              linetype = "dashed", col = "red") + 
    coord_cartesian(xlim = c(0, xlim_max)) +
    scale_x_continuous(breaks = seq(0, 1000, by = 20),
                       sec.axis = dup_axis(breaks = x$n_enriched,
                                           labels = x$NL,
                                           name = "Mouse threshold (-log10(q))")) +
    labs(x = "Number of mouse pathways enriched",
         y = "PR AUC",
         title = paste("Precision-recall pathway comparison between human", human_cluster, "and mouse", mouse_cluster)) + 
    theme_bw()
  
  # outfile <- "human_7-2_mouse_7-1_PR_AUC.pdf"
  # outfile <- file.path(output_dir, outfile)
  # pdf(file = outfile,
  #     width = unit(8, "in"),
  #     height = unit(4, "in"))
  # print(plt_H72_M71_PR_AUC)
  # dev.off()
  
  return(plt)
}

plot_enrichment_HG <- function(x, human_cluster, mouse_cluster, export = FALSE) {
  
  palette <- colorRampPalette(rev(brewer.pal(n = 9, name = "Reds")[3:9]))(10)
  
  xlim_max <- round(max(x$B), -1) + 10
  df_labs <- x %>% 
    select(alpha_mouse, B) %>%
    distinct()
  
  plt <- ggplot(x, aes(x = B, y = -log10(pval), 
                       col = factor(round(-log10(alpha_human), 2)))) + 
    geom_line() + 
    geom_point() +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
    labs(col = "Human threshold\n(-log10(q))",
         y = "-log10(p)",
         x = "Number of mouse pathways enriched",
         title = paste("Hypergeometric test pathway comparison between human", human_cluster, "and mouse", mouse_cluster)) + 
    coord_cartesian(xlim = c(0, xlim_max)) +
    scale_x_continuous(breaks = seq(0, 1000, by = 20),
                       sec.axis = dup_axis(breaks = df_labs$B,
                                           labels = round(-log10(df_labs$alpha_mouse), 2),
                                           name = "Mouse threshold (-log10(q))")) +
    # scale_y_continuous(breaks = seq(0, 20, by = 1)) + 
    scale_color_manual(values = palette) + 
    theme_bw()
  
  # outfile <- "human_7-2_mouse_7-1_HG.pdf"
  # outfile <- file.path(output_dir, outfile)
  # pdf(file = outfile,
  #     width = unit(8, "in"),
  #     height = unit(4, "in"))
  # print(plt_H72_M71_HG)
  # dev.off()
  
  return(plt)
  
}

plot_enrichment <- function(enrichment, quantiles, similarity, human_cluster, mouse_cluster) {
  
  corr <- similarity %>%
    filter(human_cluster_id == human_cluster, 
           mouse_cluster_id == mouse_cluster) %>% 
    pull(similarity) %>% 
    round(4)
  
  pval <- similarity %>% 
    filter(human_cluster_id == human_cluster, 
           mouse_cluster_id == mouse_cluster) %>% 
    pull(pval) %>% 
    round(4)
  
  cluster_stats <- paste0("Transcriptomic similarity: ", "r = ", corr, "; p = ", pval)
  
  plt_ROC <- plot_enrichment_ROC_AUC(x = enrichment[["ROC"]],
                                     quantiles = quantiles[["ROC"]],
                                     human_cluster = human_cluster,
                                     mouse_cluster = mouse_cluster) +
    labs(title = "ROC comparison")
  
  plt_PR <- plot_enrichment_PR_AUC(x = enrichment[["PR"]],
                                   quantiles = quantiles[["PR"]],
                                   human_cluster = human_cluster,
                                   mouse_cluster = mouse_cluster) +
    labs(title = "Precision-recall comparison") 
  
  plt_HG <- plot_enrichment_HG(x = enrichment[["HG"]],
                               human_cluster = human_cluster,
                               mouse_cluster = mouse_cluster) +
    labs(title = "Hypergeometric test comparison") + 
    theme(legend.position = "bottom")
  
  plt <- plt_ROC / plt_PR / plt_HG +
    plot_annotation(title = paste("Pathway correspondence between human", human_cluster, "and mouse", mouse_cluster),
                    subtitle = cluster_stats)
  
  outfile <- paste("pathway_correspondence_human", human_cluster, "mouse", mouse_cluster, sep = "_")
  outfile <- paste0(outfile, ".pdf")
  outfile <- file.path(output_dir, outfile)
  pdf(file = outfile,
      width = unit(8, "in"),
      height = unit(10, "in"))
  print(plt)
  dev.off()
  
  return(plt)
  
}
```

```{r}
# Sequence of number of pathways enriched in cluster
n_enriched_seq <- c(300, 100, 50, 25, 10, 5:1)

for (hi in 1:length(human_cluster_ids)) {
  
  # Human cluster to examine
  human_cluster_id <- human_cluster_ids[hi]
  human_nk <- str_split(human_cluster_id, "-")[[1]][1]
  human_k <- str_split(human_cluster_id, "-")[[1]][2]
  
  # Import human pathway enrichment for current cluster
  human_enrichment <- import_enrichment_human(params_id = human_params_id, 
                                              pipeline_dir = human_pipeline_dir, 
                                              nk = human_nk, k = human_k)
  
  # Filter for pathways to keep and arrange by ID for correspondence to mouse
  human_enrichment <- human_enrichment %>% 
    filter(ID %in% pathway_ids_keep) %>% 
    arrange(ID)
  
  AUC_quantiles <- get_auc_null_quantiles(enrichment = human_enrichment,
                                          n_true = n_enriched_seq)
  
  # Get mouse clusters
  idx_cluster_id <- df_matching_clusters[["human_cluster_id"]] == human_cluster_id
  mouse_cluster_ids <- df_matching_clusters[idx_cluster_id, "mouse_cluster_id"][[1]]
  mouse_cluster_ids <- c(mouse_cluster_ids, paste(human_nk, 1:as.numeric(human_nk), sep = "-"))
  mouse_cluster_ids <- unique(mouse_cluster_ids)
  
  for (mi in 1:length(mouse_cluster_ids)) {
    
    mouse_cluster_id <- mouse_cluster_ids[mi]
    mouse_nk <- str_split(mouse_cluster_id, "-")[[1]][1]
    mouse_k <- str_split(mouse_cluster_id, "-")[[1]][2]
    
    mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                                pipeline_dir = mouse_pipeline_dir, 
                                                nk = mouse_nk, k = mouse_k)
    
    mouse_enrichment <- mouse_enrichment %>% 
      filter(ID %in% pathway_ids_keep) %>% 
      arrange(ID)
    
    mouse_alpha_seq <- 2:10
    mouse_alpha_seq <- (1/10)^mouse_alpha_seq
    mouse_alpha_seq <- c(0.05, mouse_alpha_seq)
    
    enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                             human_enrichment = human_enrichment,
                                             alpha = mouse_alpha_seq)
    
    enrichment_ROC <- enrichment_ROC[!is.na(enrichment_ROC)]
    
    enrichment_ROC_AUC <- inner_join(map_dbl(enrichment_ROC, pROC::auc) %>% 
                                       enframe(name = "alpha", value = "AUC") %>% 
                                       mutate(alpha = as.numeric(alpha)),
                                     map_dbl(enrichment_ROC, function(x){sum(x[["original.response"]])}) %>% 
                                       enframe(name = "alpha", value = "n_enriched") %>% 
                                       mutate(alpha = as.numeric(alpha)),
                                     by = "alpha")
    
    enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                           human_enrichment = human_enrichment,
                                           alpha = mouse_alpha_seq)
    
    enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                                 .f = function(x){
                                   tibble(AUC = x[["auc.integral"]], 
                                          prop = x[["true.proportion"]],
                                          n_enriched = prop*nrow(human_enrichment))
                                 }, .id = "alpha") %>% 
      mutate(alpha = as.numeric(alpha)) %>% 
      filter(!is.na(AUC))
    
    enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
                                               human_enrichment = human_enrichment,
                                               alpha_mouse = mouse_alpha_seq)
    
    enrichment <- list(ROC = enrichment_ROC_AUC,
                       PR = enrichment_PR_AUC,
                       HG = enrichment_HG)
    
    out <- plot_enrichment(enrichment = enrichment,
                           quantiles = AUC_quantiles,
                           similarity = df_sim_pvals,
                           human_cluster = human_cluster_id,
                           mouse_cluster = mouse_cluster_id)
  }
}
```


### Correspondence heuristics

```{r}
matching_pairs <- df_matching_clusters %>% 
  unite(col = cluster_pair, human_cluster_id, mouse_cluster_id, remove = FALSE) %>% 
  pull(cluster_pair)

df_human_cluster_groups <- tibble(A = c("4-2", "5-2", "6-6", "7-2", "8-6", "9-2", "10-3"),
                                  B = c("4-3", "5-5", "6-2", "7-4", "8-7", "9-4", "10-4"))

list_enrichment <- vector(mode = "list", length = ncol(df_human_cluster_groups))
names(list_enrichment) <- colnames(df_human_cluster_groups)
list_quantiles <- vector(mode = "list", length = ncol(df_human_cluster_groups))
names(list_quantiles) <- colnames(df_human_cluster_groups)
for (j in 1:ncol(df_human_cluster_groups)) {
  
  human_cluster_ids <- df_human_cluster_groups[[j]]
  
  # Sequence of number of pathways enriched in cluster
  n_enriched_seq <- c(500, 400, 300, 100, 50, 25, 10, 5:1)
  
  list_enrichment_j <- vector(mode = "list", length = length(human_cluster_ids))
  names(list_enrichment_j) <- human_cluster_ids
  for (hi in 1:length(human_cluster_ids)) {
    
    # Human cluster to examine
    human_cluster_id <- human_cluster_ids[hi]
    human_nk <- str_split(human_cluster_id, "-")[[1]][1]
    human_k <- str_split(human_cluster_id, "-")[[1]][2]
    
    # Import human pathway enrichment for current cluster
    human_enrichment <- import_enrichment_human(params_id = human_params_id, 
                                                pipeline_dir = human_pipeline_dir, 
                                                nk = human_nk, k = human_k)
    
    # Filter for pathways to keep and arrange by ID for correspondence to mouse
    human_enrichment <- human_enrichment %>% 
      filter(ID %in% pathway_ids_keep) %>% 
      arrange(ID)
    
    if (hi == 1) {
      list_quantiles[[j]] <- get_auc_null_quantiles(enrichment = human_enrichment,
                                                    n_true = n_enriched_seq)
    }
    
    # Get mouse clusters
    idx_cluster_id <- df_matching_clusters[["human_cluster_id"]] == human_cluster_id
    mouse_cluster_ids <- df_matching_clusters[idx_cluster_id, "mouse_cluster_id"][[1]]
    mouse_cluster_ids <- c(mouse_cluster_ids, paste(human_nk, 1:as.numeric(human_nk), sep = "-"))
    mouse_cluster_ids <- unique(mouse_cluster_ids)
    
    list_enrichment_hi <- vector(mode = "list", length = length(mouse_cluster_ids))
    names(list_enrichment_hi) <- mouse_cluster_ids
    for (mi in 1:length(mouse_cluster_ids)) {
      
      mouse_cluster_id <- mouse_cluster_ids[mi]
      mouse_nk <- str_split(mouse_cluster_id, "-")[[1]][1]
      mouse_k <- str_split(mouse_cluster_id, "-")[[1]][2]
      
      mouse_enrichment <- import_enrichment_mouse(params_id = mouse_params_id, 
                                                  pipeline_dir = mouse_pipeline_dir, 
                                                  nk = mouse_nk, k = mouse_k)
      
      mouse_enrichment <- mouse_enrichment %>% 
        filter(ID %in% pathway_ids_keep) %>% 
        arrange(ID)
      
      mouse_alpha_seq <- 2:10
      mouse_alpha_seq <- (1/10)^mouse_alpha_seq
      mouse_alpha_seq <- c(0.05, mouse_alpha_seq)
      
      enrichment_ROC <- compute_enrichment_ROC(mouse_enrichment = mouse_enrichment, 
                                               human_enrichment = human_enrichment,
                                               alpha = mouse_alpha_seq)
      
      enrichment_ROC <- enrichment_ROC[!is.na(enrichment_ROC)]
      
      enrichment_ROC_AUC <- inner_join(map_dbl(enrichment_ROC, pROC::auc) %>% 
                                         enframe(name = "alpha", value = "AUC") %>% 
                                         mutate(alpha = as.numeric(alpha)),
                                       map_dbl(enrichment_ROC, function(x){sum(x[["original.response"]])}) %>% 
                                         enframe(name = "alpha", value = "n_enriched") %>% 
                                         mutate(alpha = as.numeric(alpha)),
                                       by = "alpha")
      
      enrichment_PR <- compute_enrichment_PR(mouse_enrichment = mouse_enrichment,
                                             human_enrichment = human_enrichment,
                                             alpha = mouse_alpha_seq)
      
      enrichment_PR_AUC <- map_dfr(.x = enrichment_PR, 
                                   .f = function(x){
                                     tibble(AUC = x[["auc.integral"]], 
                                            prop = x[["true.proportion"]],
                                            n_enriched = prop*nrow(human_enrichment))
                                   }, .id = "alpha") %>% 
        mutate(alpha = as.numeric(alpha)) %>% 
        filter(!is.na(AUC))
      
      # enrichment_HG <- compute_enrichment_HGtest(mouse_enrichment = mouse_enrichment, 
      #                                            human_enrichment = human_enrichment,
      #                                            alpha_mouse = mouse_alpha_seq)
      # 
      
      list_enrichment_hi[[mi]] <- list(ROC = enrichment_ROC_AUC,
                                       PR = enrichment_PR_AUC)
      
    }
    
    list_enrichment_j[[hi]] <- list(ROC = map_dfr(list_enrichment_hi,
                                                  function(x){x[["ROC"]]}, 
                                                  .id = "mouse_cluster_id"),
                                    PR = map_dfr(list_enrichment_hi,
                                                 function(x){x[["PR"]]}, 
                                                 .id = "mouse_cluster_id"))
  }
  
  df_ROC_j <- map_dfr(list_enrichment_j, function(x){x[["ROC"]]}, .id = "human_cluster_id") %>% 
    unite(col = cluster_pair, human_cluster_id, mouse_cluster_id, remove = FALSE) %>% 
    mutate(matching = cluster_pair %in% matching_pairs,
           NL = round(-log10(alpha), 2))
  
  df_PR_j <- map_dfr(list_enrichment_j, function(x){x[["PR"]]}, .id = "human_cluster_id") %>% 
    unite(col = cluster_pair, human_cluster_id, mouse_cluster_id, remove = FALSE) %>% 
    mutate(matching = cluster_pair %in% matching_pairs,
           NL = round(-log10(alpha), 2))
  
  list_enrichment[[j]] <- list(ROC = df_ROC_j,
                               PR = df_PR_j)
  
}
```


```{r}
for (j in 1:length(list_enrichment)) {
  
  df_ROC <- list_enrichment[[j]][["ROC"]]
  df_PR <- list_enrichment[[j]][["PR"]]
  
  plt_ROC <- ggplot(df_ROC, aes(x = n_enriched, y = AUC, group = cluster_pair, col = matching)) + 
    geom_line(linewidth = 0.3, alpha = 0.5) + 
    geom_point(size = 1) +
    geom_ribbon(data = list_quantiles[[j]]$ROC, 
                mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
                inherit.aes = FALSE,
                alpha = 0.1, fill = "blue") + 
    geom_hline(yintercept = 0.5, linetype = "dashed", col = "blue") + 
    coord_cartesian(xlim = c(0, 400),
                    ylim = c(0.4, 1.0)) +
    scale_x_continuous(breaks = seq(0, 1000, by = 50)) + 
    scale_y_continuous(breaks = seq(0, 1, by = 0.1)) + 
    scale_color_manual(values = c("grey50", "red"), labels = c("No", "Yes")) + 
    labs(x = "Number of mouse pathways enriched",
         y = "ROC AUC",
         col = "Matching pair") + 
    theme_bw()
  
  plt_PR <- ggplot(df_PR, aes(x = n_enriched, y = AUC, group = cluster_pair, col = matching)) + 
    geom_line(linewidth = 0.3, alpha = 0.5) + 
    geom_point(size = 1) +
    geom_ribbon(data = list_quantiles[[j]]$PR, 
                mapping = aes(x = n_enriched, ymin = AUC_lwr, ymax = AUC_upr),
                inherit.aes = FALSE,
                alpha = 0.1, fill = "blue") + 
    geom_line(data = list_quantiles[[j]]$PR, 
              mapping = aes(x = n_enriched, y = AUC_chance),
              inherit.aes = FALSE,
              linetype = "dashed", col = "blue") + 
    coord_cartesian(xlim = c(0, 400)) +
    scale_x_continuous(breaks = seq(0, 1000, by = 50)) +
    scale_color_manual(values = c("grey50", "red"), labels = c("No", "Yes")) + 
    labs(x = "Number of mouse pathways enriched",
         y = "PR AUC",
         col = "Matching pair") + 
    theme_bw()
  
  plt <- plt_ROC / plt_PR + plot_annotation(title = paste("Pathway enrichment correspondence for human group", names(list_enrichment)[j])) + plot_layout(guides = "collect")
  
  outfile <- paste("pathway_correspondence_group", names(list_enrichment)[j], sep = "_")
  outfile <- paste0(outfile, ".pdf")
  outfile <- file.path(output_dir, outfile)
  pdf(file = outfile,
      width = unit(8, "in"),
      height = unit(10, "in"))
  print(plt)
  dev.off()
  
}
```


```{r}
df_human_cluster_groups_ngenes <- df_human_cluster_groups %>% 
  pivot_longer(cols = everything(), names_to = "group", values_to = "cluster_id") %>% 
  left_join(df_human_cluster_ngenes, by = "cluster_id")

ggplot(df_human_cluster_groups_ngenes, 
       aes(x = nk, y = n, col = group)) + 
  geom_line() + 
  geom_point() + 
  coord_cartesian(xlim = c(3, 10),
                  ylim = c(3, 7)) + 
  scale_x_continuous(breaks = seq(0, 10, 1)) + 
  scale_y_continuous(breaks = seq(0, 10, 1)) + 
  labs(x = "nk",
       y = "Number of genes",
       col = "Cluster group") + 
  theme_bw()
```

```{r}
df_human_cluster_variants_long <- df_human_variants %>% 
  select(Subject_ID, contains("nk"), gene) %>% 
  pivot_longer(cols = contains("nk"), names_to = "nk_name", values_to = "k") %>% 
  mutate(nk = str_remove(nk_name, "nk")) %>% 
  select(-nk_name) %>% 
  unite(col = "cluster_id", nk, k, sep = "-", remove = FALSE)

df_human_cluster_variants_long %>% 
  filter(cluster_id %in% df_human_cluster_groups$A,
         !is.na(gene)) %>% 
  pull(gene) %>% unique()

df_human_cluster_variants_long %>% 
  filter(cluster_id %in% df_human_cluster_groups$B,
         !is.na(gene)) %>% 
  pull(gene) %>% unique()
```

```{r}
df_human_group_A_ranks <- readxl::read_excel(file.path(output_dir, "human_cluster_pathway_ranks.xlsx"))

df_human_group_A_ranks <- df_human_group_A_ranks %>% 
  group_by(human_cluster_id) %>% 
  mutate(norm_rank = (rank - 1)/(n() - 1)) %>% 
  ungroup() 

ranks_matching <- df_human_group_A_ranks %>% 
  filter(match == 1) %>% 
  pull(norm_rank)

wilcox.test(ranks_matching, mu = 0.5, alternative = "less")
```

```{r}
human_cluster_ids <- df_human_group_A_ranks %>% 
  filter(match == 1) %>% 
  pull(human_cluster_id)

n_perm <- 1e4
mat_sample_ranks <- matrix(data = 0, nrow = n_perm, ncol = length(human_cluster_ids))
colnames(mat_sample_ranks) <- human_cluster_ids

for (i in 1:n_perm) {
  set.seed(i)
  mat_sample_ranks[i,] <-
    df_human_group_A_ranks %>% 
    select(human_cluster_id, mouse_cluster_id, match) %>% 
    group_by(human_cluster_id) %>% 
    mutate(rank = sample(1:n(), size = n(), replace = FALSE),
           norm_rank = (rank - 1)/(n() - 1)) %>% 
    ungroup() %>% 
    filter(match == 1) %>% 
    pull(norm_rank) 
}

sample_rank_means <- rowSums(mat_sample_ranks)/ncol(mat_sample_ranks)

true_rank_mean <- df_human_group_A_ranks %>% 
  filter(match == 1) %>% 
  pull(norm_rank) %>% 
  mean()

sum(sample_rank_means <= true_rank_mean)/n_perm
```

```{r}
ggplot(tibble(x = sample_rank_means),
       aes(x = x)) + 
  geom_histogram(fill = "grey80",
                 col = "grey20",
                 bins = 30) +
  annotate(geom = "segment",
           x = true_rank_mean, xend = true_rank_mean,
           y = 0, yend = 2000,
           linetype = "dashed", col = "red") + 
  coord_cartesian(ylim = c(0, 1000)) + 
  labs(x = "Normalized rank",
       y = "n") + 
  theme_bw()
```





```{r}
data.frame(instance = c(rep(1, 4), rep(2, 5), rep(3, 6), rep(4, 7), rep(5, 8)),
           rank = c(1:4, 1:5, 1:6, 1:7, 1:8),
           group = c("B", "A", "A", "A", 
                     "A", "B", "A", "A", "A",
                     "A", "B", "A", "A", "A", "A",
                     "A", "B", "A", "A", "A", "A", "A",
                     "A", "B", "A", "A", "A", "A", "A","A"))
```



# Development


